### PR TITLE
[OPT Serving] Support batched inference

### PR DESCRIPTION
**Pull request by zhisbug**

### CODE REVIEW

**Comment by merrymercy** [comment]
Let me know when you fix all hard-coded values and the PR is ready for review

**Comment by zhisbug** [comment]
@merrymercy all bugs fixed

File: examples/opt_serving/model/opt_model.py [hunk 1]
```diff
@@ -175,11 +176,14 @@ def __call__(self,
             mask = jnp.arange(max_length) - (cache_index_ + 1)
             attention_bias = jnp.expand_dims(
                 (row_idxs[:, None] <= mask).astype(self.dtype) * -1e10, (0, 1))
+            # if attention_mask is not None:
+            #     attention_bias += attention_mask
```
Comment (merrymercy) [inline examples/opt_serving/model/opt_model.py:180]: remove comments

File: examples/opt_serving/model/wrapper.py [hunk 2]
```diff
@@ -184,6 +191,30 @@ def grouped_reorder_cache(arys, device_mesh):
...
+        if isinstance(attention_mask, torch.Tensor):
+            attention_mask = attention_mask.cpu().numpy()
+
+        # attention_mask =  attention_mask == 1
+        batch_size = attention_mask.shape[0]
+        # # flip
+        # big_neg = jnp.finfo(dtype).min
+        # flipped = (attention_mask == 0) * big_neg
+        # # extend
+        ret_mask = np.ones((batch_size, self._max_target_positions),
```
Comment (merrymercy) [inline examples/opt_serving/model/wrapper.py:210]: one or zero?

File: examples/opt_serving/model/wrapper.py [hunk 3]
```diff
@@ -184,6 +191,30 @@ def grouped_reorder_cache(arys, device_mesh):
...
+    def _process_attention_mask(self, attention_mask):
+        if isinstance(attention_mask, torch.Tensor):
+            attention_mask = attention_mask.cpu().numpy()
+
+        # attention_mask =  attention_mask == 1
+        batch_size = attention_mask.shape[0]
+        # # flip
+        # big_neg = jnp.finfo(dtype).min
+        # flipped = (attention_mask == 0) * big_neg
+        # # extend
```
Comment (merrymercy) [inline examples/opt_serving/model/wrapper.py:209]: remove comments

File: examples/opt_serving/textgen_demo.py [hunk 4]
```diff
@@ -1,27 +1,35 @@
...
 from opt_serving.model.wrapper import get_model
+import numpy as np
 
 # Load the tokenizer. We have to use the 30B version because
 # other versions have some issues. The 30B version works for all OPT models.
 tokenizer = AutoTokenizer.from_pretrained("facebook/opt-30b", use_fast=False)
 tokenizer.add_bos_token = False
 
-generate_params = {"do_sample": True, "num_beams": 4, "num_return_sequences": 3}
+generate_params = {"do_sample": False, "num_beams": 1, "num_return_sequences": 1}
```
Comment (merrymercy) [inline examples/opt_serving/textgen_demo.py:11]: If batch compatible with beam search?
If not, where should we raise errors?

**Review by merrymercy — CHANGES_REQUESTED** [review]

File: examples/opt_serving/model/wrapper.py [hunk 5]
```diff
@@ -219,14 +250,22 @@ def get_hf_opt_model(model_name, device, num_beams):
...
             past_length = past_key_values[0][0].shape[2]
-            attention_mask = torch.ones(
-                (input_ids.shape[0], past_length + 1)).to(device)
+        if past_length >= attention_length:
+            attention_mask = torch.cat(
+                (attention_mask,
+                 torch.ones((input_ids.shape[0],
+                             past_length + 1 - attention_length)).to(device)),
+                dim=1)
+        elif attention_mask != None:
```
Comment (zhuohan123) [inline examples/opt_serving/model/wrapper.py:267]: `attention_mask` will never be `None` here since you take the size of it on line 256.

File: examples/opt_serving/model/wrapper.py [hunk 6]
```diff
@@ -219,14 +250,22 @@ def get_hf_opt_model(model_name, device, num_beams):
...
                        output_hidden_states=False):
+        attention_length = attention_mask.size()[-1]
         if past_key_values is None:
-            attention_mask = None
+            past_length = 0
         else:
             past_length = past_key_values[0][0].shape[2]
-            attention_mask = torch.ones(
-                (input_ids.shape[0], past_length + 1)).to(device)
+        if past_length >= attention_length:
```
Comment (zhuohan123) [inline examples/opt_serving/model/wrapper.py:261]: Question: Are there cases that `past_length != attention_length`?

**Review by zhuohan123 — CHANGES_REQUESTED** [review]

File: examples/opt_serving/textgen_demo.py (hunk 4 above)
Comment (vinlnx) [reply examples/opt_serving/textgen_demo.py:11]: It should be compatible with beam search.

File: examples/opt_serving/model/wrapper.py (hunk 5 above)
Comment (zhisbug) [reply examples/opt_serving/model/wrapper.py:267]: @ZYHowell

File: examples/opt_serving/model/wrapper.py (hunk 2 above)
Comment (zhisbug) [reply examples/opt_serving/model/wrapper.py:210]: does not matter because positions that are autoregressive generated won't be impacted by masks.

File: examples/opt_serving/textgen_demo.py (hunk 4 above)
Comment (zhisbug) [reply examples/opt_serving/textgen_demo.py:11]: it works fine.
